### PR TITLE
Add judge metric

### PR DESCRIPTION
**Pull request by perlitz**
Adding a judge metric example

### CODE REVIEW

File: src/unitxt/metrics.py [hunk 1]
```diff
@@ -1387,6 +1392,98 @@ def compute(
...
+        metric = LlamaIndexCorrectnessMetric()
+        metric.prepare()
+        scores = metric.compute(references, predictions, additional_inputs)
+    """
+
+    reduction_map = {"mean": ["score"]}
+    main_score = "score"
+    # model_name: str
+
+    _requirements_list: List[str] = []
```
Comment (elronbandel) [inline src/unitxt/metrics.py:1417]:
```suggestion
    _requirements_list: List[str] = ["llama_index"]
```

Comment (elronbandel) [inline src/unitxt/metrics.py:1417]:
```suggestion
    _requirements_list: List[str] = ["llama_index"]
```

File: src/unitxt/metrics.py [hunk 2]
```diff
@@ -1387,6 +1392,98 @@ def compute(
...
+        main_score (str): The main score used for evaluation.
+        _requirements_list (List[str]): A list specifying any additional requirements for the metric.
+
+    Methods:
+        prepare(self): Initialization method for the metric.
+        compute(self, references, predictions, additional_inputs): Method to compute the metric.
+
+    Usage:
+        metric = LlamaIndexCorrectnessMetric()
+        metric.prepare()
```
Comment (elronbandel) [inline src/unitxt/metrics.py:1409]:
```suggestion
```
happens automatically

File: prepare/metrics/llama_index_corectness.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#     metric=LlamaIndexCorrectnessMetric(),
+#     postpreprocess_steps=[
+#         # CopyFields(
+#         #     field_to_field=[
+#         #         ("score/instance/score", "score"),
+#         #     ],
+#         #     use_query=True,
+#         # )
+#     ],
+# )
```
Comment (elronbandel) [inline prepare/metrics/llama_index_corectness.py:24]:
```suggestion
```
why this endded up not working? Just curious

File: src/unitxt/metrics.py [hunk 4]
```diff
@@ -1265,6 +1266,10 @@ class TokenOverlap(InstanceMetric):
     def compute(
         self, references: List[Any], prediction: Any, additional_inputs: List[Dict]
     ) -> dict:
+        import ast
+
+        references = ast.literal_eval(references)
```
Comment (yoavkatz) [inline src/unitxt/metrics.py:1271]: Why is this needed?

File: prepare/metrics/llama_index_corectness.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
+from unitxt import add_to_catalog
+from unitxt.metrics import LlamaIndexCorrectnessMetric
+from unitxt.test_utils.metrics import test_metric
```
Comment (yoavkatz) [inline prepare/metrics/llama_index_corectness.py:3]: rename file from corectness-> correctness

File: src/unitxt/metrics.py [hunk 6]
```diff
@@ -1387,6 +1392,98 @@ def compute(
         return self.pipe(inputs, batch_size=self.batch_size)
 
 
+class LlamaIndexCorrectnessMetric(BulkInstanceMetric):
+    """Custom metric class for evaluating correctness using Llama Index.
+
```
Comment (yoavkatz) [inline src/unitxt/metrics.py:1397]: It requires OpenAI API .  How is it specified?

File: prepare/metrics/llama_index_corectness.py (hunk 3 above)
Comment (dafnapension) [reply prepare/metrics/llama_index_corectness.py:24]: `ast.literal_eval(references)`   where `references` is a list of two strings. but `literal_eval` expects `node_or_string`

File: src/unitxt/metrics.py (hunk 4 above)
Comment (perlitz) [reply src/unitxt/metrics.py:1271]: I might be doing something wrong but references were definitely a string 
```python
"['reference1',...]"
```

File: prepare/metrics/llama_index_corectness.py (hunk 5 above)
Comment (perlitz) [reply prepare/metrics/llama_index_corectness.py:3]: sure

File: src/unitxt/metrics.py (hunk 6 above)
Comment (perlitz) [reply src/unitxt/metrics.py:1397]: Using an API key in an environmental variable.

BTW, this is something we need to be careful about (or even consider having) as proprietary data can be leaked this way to OPENAI

Generally, this is a major issue with some packages like llama-index as they sometimes use OPENAI based models as default.

Comment (yoavkatz) [reply src/unitxt/metrics.py:1397]: I think the 

class LlamaIndexCorrectnessMetric(BulkInstanceMetric):

should br sn abstract class with method get_llm() .

Then there should be a class 

class LlamaIndexCorrectnessMetricByOpenAI(LlamaIndexCorrectnessMetric):

which would receive a parameter (model name, could be with default "gpt-3.5-turbo")
 
 and the get_llm method of it would be
 
  return OpenAI(mode_name)

Comment (yoavkatz) [reply src/unitxt/metrics.py:1397]: Then the name in the catalog would also be

correctness_by_lamma_index_openai

Then it will be clear which metric is used, and that it's with open.

Comment (yoavkatz) [reply src/unitxt/metrics.py:1397]: Also, I assume this should be the name of the returned score.

File: src/unitxt/metrics.py (hunk 4 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:1271]: I think the issue is that RAG task you defined as a field answers (which is a list of possible answers), but you used InputOutputTemplate and not MultiReferenceTemplate.  So the list of answers was converted to a string representation.

@elronbandel - we need to better deal with this.

Comment (perlitz) [reply src/unitxt/metrics.py:1271]: even when I am using MultiReferenceTemplate I have the same problem
note that when I test the metric I am not facing it, only when I test it within the card testing
